JSSpecCompiler: Road to testing (part 3 out of 2)

DanShaders opened the pull request:
This actually adds a test!

ADKaster commented on Tests/JSSpecCompiler/test-runner.cpp:
```diff
@@ -0,0 +1,184 @@
+
+TEST_CASE(test_regression)
+{
+    auto* source_dir_env = getenv("SERENITY_SOURCE_DIR");
+    auto* binary_dir_env = getenv("LAGOM_BINARY_DIR");
+
+    DeprecatedString source_dir;
+    DeprecatedString binary_dir;
+
+    if (source_dir_env == nullptr || binary_dir_env == nullptr) {
+        warnln("SERENITY_SOURCE_DIR or LAGOM_BINARY_DIR is not set. Assuming . for SERENITY_SOURCE_DIR and ./Build/lagom/bin for LAGOM_BINARY_DIR.");
+        source_dir = MUST(Core::System::getcwd());
+        binary_dir = LexicalPath::join(source_dir, "Build/lagom/bin"sv).string();
```
These assumptions are kinda bunk. We don't want to assume Build/Lagom/bin.

For `path_to_compiler_binary`, you want something more like

```
auto path_to_self = LexicalPath(MUST(Core::System::current_executable_path())).parent();
auto path_to_compiler_binary = LexicalPath::join(path_to_self, compiler_binary);
```

Then we can yeet LAGOM_BINARY_DIR, we don't need it. 

For SERENITY_SOURCE_DIR, we should get rid of that as well, and instead assume that the tests will be in $PWD/Tests or some such. then you can set the working directory of the test runner to Tests/JSSpecCompiler/ and reference them from Core::System::cwd() or whatever we called that.

I'm not a fan of these build system specifics leaking into the binaries. ideally they should be relocatable. If you want we could change the way TestSuite::main() calls m_setup to let it take the Main::Arguments to parse extra ones. or take the ArgsParser by reference before calling parse(). That way when run in CI or via the scripts, we can assume the "standard paths", and you can also pass in --test-dir or --compiler-path if you like to the binary to run it manually.

These steps here should be in TEST_SETUP() and stored in some file scope globals rather than in the TEST_CASE, imo.

ADKaster commented on Tests/JSSpecCompiler/test-runner.cpp:
```diff
@@ -0,0 +1,184 @@
+
+            auto arguments = build_command_line_arguments(path_to_compiler_binary, path_to_test, test_description);
+
+            // FIXME: We should invent better API to invoke a helper process.
```
We have Core::Command, Core::Process::spawn(), and whatever `Utilities/run-tests.cpp` does. was this copied from one of those?

DanShaders replied:
Not really, I've written this myself :) `run-tests.cpp` seem to also use manual `posix_spawn`.
I guess I will then just add support for file actions to `Core::Process` and use it here and make a separate PR to port `run-tests.cpp` to it.

DanShaders commented:
Sorry for force-push spam, but this should be good now.

Changes since review:
- Removed path assumptions based on build system specifics.
- Stopped forwarding `SERENITY_SOURCE_DIR` and `LAGOM_BINARY_DIR`.
- Unported System::current_executable_path from new String (since it was trying to store a path in it).
- Refactored `Core::Process` to be a little bit more like `Core::File` and to support file actions.
- And used it in test runner.

ADKaster commented on Userland/Libraries/LibCore/Process.cpp:
```diff
@@ -57,64 +56,105 @@ struct ArgvList {
-        auto pid = TRY(System::posix_spawn(m_path.view(), nullptr, nullptr, const_cast<char**>(get().data()), System::environment()));
-        // FIXME: Support keep_as_child outside Serenity.
-        (void)keep_as_child;
+        TODO();
```
Feels like it could do with a comment of what we actually want to do here. Something like `// FIXME: Support chdir file options on platforms that support it`

ADKaster commented on Userland/Libraries/LibCore/Process.cpp:
```diff
@@ -57,64 +56,105 @@ struct ArgvList {
+    }));
+
+    if (keep_as_child == KeepAsChild::No)
+        TRY(process.disown());
```
Food for thought: are we actually using disown anywhere in the system? From what I remember it was an experimental thing from way early in development that isn't really a thing on any other unix.

DanShaders replied:
`KeepAsChild::No` is a default option for `keep_as_child`, so it is implicitly used in a lot of places. That said, the only place where we pass `KeepAsChild::No` explicitly is in SystemServer when it spawns `/bin/DeviceMapper`.

ADKaster commented on Userland/Libraries/LibCore/Process.cpp:
```diff
@@ -57,64 +56,105 @@ struct ArgvList {
-    }
-    argv.set_working_directory(working_directory);
-    return argv.spawn(keep_as_child);
+    for (auto const& argument : arguments)
```
This refactor skips the ensure_capacity for the backing strings vector.

ADKaster commented on Userland/Libraries/LibCore/Process.cpp:
```diff
@@ -260,4 +300,40 @@ void Process::wait_for_debugger_and_break()
+        exited_with_code_0 &= WEXITSTATUS(status) == 0;
+        dbgln("here with status {}", WEXITSTATUS(status));
+    } else {
+        // FIXME: Did it really exit, though?
```
ehm. This should call WIFSIGNALED and WIFSTOPPED as well for completeness

ADKaster commented on Userland/Libraries/LibCore/Process.h:
```diff
@@ -8,20 +8,62 @@
+    DeprecatedString path;
+    File::OpenMode mode;
+    int fd;
+    int permissions = 0600;
```
Do we not have a permission mode enum for files elsewhere in LibCore? if not we should probably have one in the future.

DanShaders replied:
We don't. I can change the type from `int` to `mode_t`, though.

ADKaster commented on Userland/Libraries/LibCore/Process.h:
```diff
@@ -8,20 +8,62 @@
+struct OpenFile {
+    DeprecatedString path;
+    File::OpenMode mode;
+    int fd;
```
`fd` should be default init to -1. and open mode to ... something? Gotta explicitly default init all those primitive types.
